### PR TITLE
Allow blocks to be empty

### PR DESCRIPTION
**Pull request by gzanitti**
This PR builds on top of #1393.
It mainly modifies the grammar by changing the way SelectedExpressions are declared, to allow blocks to be empty.

### CODE REVIEW

**Comment by chriseth** [comment]
Is this ready? (Asking because it's in draft mode)

**Comment by gzanitti** [comment]
> Is this ready? (Asking because it's in draft mode)

Sorry, I forgot to change the status when we merge the other PR on empty blocks. It's ready now.

File: analysis/README.md [hunk 1]
```diff
@@ -6,17 +6,17 @@ This crate is where most of the compilation pipeline lives. It takes a parse tre
 
 We define two types of machines: virtual machines and constrained machines. Constrained machines are the lower level kind of machine. They have a notion of blocks through a latch and an operation_id. Virtual machines are a higher level type of machines. For each machine type, we provide the elements which appear in that type.
 
-|                         | Virtual | Constrained |
```
Comment (chriseth) [inline analysis/README.md:9]: Please undo this change.

File: analysis/README.md [hunk 2]
```diff
@@ -99,6 +99,7 @@ ASM to PIL has two steps: ROM generation and reduction to constrained.
 ##### ROM generation
 
 Rom generation generates a single ROM for each virtual machine using the following process:
+
```
Comment (chriseth) [inline analysis/README.md:102]: Same with all the newline changes here - I assume they were by accident.

File: ast/src/analyzed/display.rs [hunk 3]
```diff
@@ -267,7 +267,35 @@ impl Display for RepeatedArray {
     }
 }
 
-impl Display for Identity<Expression> {
+impl Display for parsed::SelectedExpressions<Expression> {
```
Comment (chriseth) [inline ast/src/analyzed/display.rs:270]: Shouldn't this be in src/parsed/display.rs?

File: ast/src/analyzed/mod.rs [hunk 4]
```diff
@@ -625,21 +628,51 @@ impl PublicDeclaration {
     }
 }
 
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Serialize, Deserialize, JsonSchema)]
+pub struct SelectedExpressions<Expr> {
+    pub selector: Option<Expr>,
+    pub expressions: Vec<Expr>,
+}
+
+impl<Expr> Default for SelectedExpressions<Expr> {
```
Comment (chriseth) [inline ast/src/analyzed/mod.rs:637]: Isn't this the derived implementation?

Comment (chriseth) [reply ast/src/analyzed/mod.rs:637]: Oh, it doesn't work because Expr doesn't need to implement Default, right?

File: analysis/README.md (hunk 2 above)
Comment (gzanitti) [reply analysis/README.md:102]: Yes, probably the autoformatting when saving did its thing. I'll fix it ASAP

File: ast/src/analyzed/mod.rs [hunk 5]
```diff
@@ -625,21 +628,51 @@ impl PublicDeclaration {
...
     /// For a simple polynomial identity, the selector contains
     /// the actual expression (see expression_for_poly_id).
-    pub left: SelectedExpressions<Expr>,
-    pub right: SelectedExpressions<Expr>,
+    pub left: SelectedExpressions,
+    pub right: SelectedExpressions,
 }
 
-impl<Expr> Identity<Expr> {
+impl<T> Identity<SelectedExpressions<AlgebraicExpression<T>>> {
```
Comment (chriseth) [inline ast/src/analyzed/mod.rs:669]: Why not for generic `Expr`, i.e. `SelectedExpressions<Expr>`?

File: book/src/pil/types.md [hunk 6]
```diff
@@ -241,13 +240,15 @@ References to columns have type `expr` and `expr` also implements `FromLiteral`,
 which means that literal numbers can be used in contexts where `expr` is expected.
 
 Example:
+
```
Comment (chriseth) [inline book/src/pil/types.md:243]: In this files are also some auto-changes.

File: executor/src/witgen/global_constraints.rs [hunk 7]
```diff
@@ -102,15 +102,17 @@ impl<T: FieldElement> RangeConstraintSet<&AlgebraicReference, T> for GlobalConst
     }
 }
 
+type SelExpressions<T> = SelectedExpressions<Expression<T>>;
```
Comment (chriseth) [inline executor/src/witgen/global_constraints.rs:105]: I think in witgen, there should be something like
`type Identity<T> = Identity<SelectedExpressions<Expression<T>>>`

Comment (chriseth) [reply executor/src/witgen/global_constraints.rs:105]: Maybe this even makes sense in some parts of analyzed...

File: parser/src/powdr.lalrpop [hunk 8]
```diff
@@ -622,19 +621,16 @@ MatchArm: MatchArm = {
...
 
-BlockExpression: Box<Expression> = {
-    <start:@L> "{" <statements:StatementInsideBlock+> <expr:BoxedExpression> "}" <end:@R> => ctx.to_expr_with_source_ref(BlockExpression{statements, expr}, start, end),
-}
 
-BracedExpression: Box<Expression> = {
-    <start:@L> "{" <statements:StatementInsideBlock+> <expr:BoxedExpression> "}" <end:@R> => ctx.to_expr_with_source_ref(BlockExpression{statements, expr}, start, end),
-    <start:@L> "{" <expr:BoxedExpression> "}" <end:@R> => ctx.to_expr_with_source_ref(BlockExpression{statements: vec![], expr}, start, end),
+BlockExpression: Box<Expression> = {
+    <start:@L> "{" <statements:StatementInsideBlock*> "}" <end:@R> => ctx.to_expr_with_source_ref(BlockExpression{statements, expr: Box::new(Expression::Tuple(ctx.source_ref(start, end), vec![]))}, start, end),
```
Comment (chriseth) [inline parser/src/powdr.lalrpop:632]: I'm not sure I would implement the tuple at the syntax level. Is it much more difficult to make `expr` and `Option<Expression>`?

Comment (chriseth) [reply parser/src/powdr.lalrpop:632]: Also the source reference is misleading.

File: pil-analyzer/src/condenser.rs [hunk 9]
```diff
@@ -280,18 +283,14 @@ impl<'a, T: FieldElement> Condenser<'a, T> {
 
     fn condense_selected_expressions(
         &mut self,
-        sel_expr: &'a SelectedExpressions<Expression>,
-    ) -> SelectedExpressions<AlgebraicExpression<T>> {
-        SelectedExpressions {
+        sel_expr: &'a powdr_ast::parsed::SelectedExpressions<Expression>,
+    ) -> powdr_ast::analyzed::SelectedExpressions<AlgebraicExpression<T>> {
```
Comment (chriseth) [inline pil-analyzer/src/condenser.rs:287]: is the `powdr_ast` prefix needed?

File: pil-analyzer/src/expression_processor.rs [hunk 10]
```diff
@@ -84,6 +84,23 @@ impl<'a, D: AnalysisDriver> ExpressionProcessor<'a, D> {
             .collect()
     }
 
+    pub fn process_vec_into_array_literal(&mut self, exprs: Vec<parsed::Expression>) -> Expression {
+        let unknown = SourceRef::unknown();
+        let src = exprs
+            .first()
+            .map(|e| e.source_reference())
```
Comment (chriseth) [inline pil-analyzer/src/expression_processor.rs:91]: Could you make the start the min of all starts and the end the max of all ends?

File: pil-analyzer/src/statement_processor.rs [hunk 11]
```diff
@@ -340,17 +340,19 @@ where
...
                 IdentityKind::Connect,
-                SelectedExpressions {
+                powdr_ast::parsed::SelectedExpressions {
                     selector: None,
-                    expressions: self
-                        .expression_processor(&Default::default())
-                        .process_expressions(left),
+                    expressions: Box::new(
+                        self.expression_processor(&Default::default())
+                            .process_vec_into_array_literal(left),
```
Comment (chriseth) [inline pil-analyzer/src/statement_processor.rs:347]: Is there a way to handle connect in the same way as `in`? I know it doesn't have a selector but that's the only difference, right?

File: pil-analyzer/src/pil_analyzer.rs [hunk 12]
```diff
@@ -287,8 +288,16 @@ impl PILAnalyzer {
                     if let Some(selector) = &mut part.selector {
                         expressions.push((selector, Type::Expr.into()))
                     }
-                    for e in &mut part.expressions {
-                        expressions.push((e, Type::Expr.into()))
+                    if let Expression::ArrayLiteral(_, ArrayLiteral { items }) =
```
Comment (chriseth) [inline pil-analyzer/src/pil_analyzer.rs:291]: It's fine to use `x $ f() in y $ g()` now, as long as `f()` and `g()` have the type `Expr[]`. So you shouldn't iterate through the items here but just request an array as type for `part.expressions`.

File: ast/src/analyzed/mod.rs (hunk 5 above)
Comment (gzanitti) [reply ast/src/analyzed/mod.rs:669]: The problem here is that now there are two SelectedExpression (`parsed` and `analyzed`) and in some cases functions specific to SelectedExpressions, such as `contains_next_ref`, are call, so you end up with a lot more duplicate code.

Although maybe a bit "uglier", I think this solution is cleaner.

File: ast/src/analyzed/mod.rs (hunk 4 above)
Comment (gzanitti) [reply ast/src/analyzed/mod.rs:637]: It worked, but it was only used in the right SelectedExpression in the Identity impl. I found it more practical to delete this implementation and hardcode the default expression. Thanks!

File: backend/src/estark/json_exporter/mod.rs [hunk 13]
```diff
@@ -508,93 +508,93 @@ mod test {
...
-        }
-    }
-
-    fn compare_export_file_ignore_idq_hex(file: &str) {
-        let (mut json_out, mut pilcom_parsed) = generate_json_pair(file);
-        normalize_idq_and_hex(&mut json_out);
-        normalize_idq_and_hex(&mut pilcom_parsed);
-        assert_eq!(json_out, pilcom_parsed);
-    }
+    // fn normalize_idq_and_hex(v: &mut JsonValue) {
```
Comment (chriseth) [inline backend/src/estark/json_exporter/mod.rs:511]: Let's just remove these tests.

File: executor/src/witgen/generator.rs [hunk 14]
```diff
@@ -24,9 +24,9 @@ struct ProcessResult<'a, T: FieldElement> {
 }
 
 pub struct Generator<'a, T: FieldElement> {
-    connecting_identities: BTreeMap<u64, &'a Identity<Expression<T>>>,
+    connecting_identities: BTreeMap<u64, &'a Identity<SelectedExpressions<Expression<T>>>>,
```
Comment (chriseth) [inline executor/src/witgen/generator.rs:27]: Same in this file.

File: executor/src/witgen/global_constraints.rs [hunk 15]
```diff
@@ -102,15 +103,17 @@ impl<T: FieldElement> RangeConstraintSet<&AlgebraicReference, T> for GlobalConst
     }
 }
 
+type Identity<T> = AnalyzedIdentity<SelectedExpressions<Expression<T>>>;
```
Comment (chriseth) [inline executor/src/witgen/global_constraints.rs:106]: Please move this to a more central point. Could be at the root of this crate or even at some exeternal-facing point of the podr_ast::analyzed module.

File: executor/src/witgen/identity_processor.rs [hunk 16]
```diff
@@ -126,7 +127,7 @@ impl<'a, 'b, 'c, T: FieldElement, Q: QueryCallback<T>> IdentityProcessor<'a, 'b,
     /// Returns the updates.
     pub fn process_identity(
         &mut self,
-        identity: &'a Identity<Expression<T>>,
+        identity: &'a Identity<SelectedExpressions<Expression<T>>>,
```
Comment (chriseth) [inline executor/src/witgen/identity_processor.rs:130]: Should use the typedef as well.

File: executor/src/witgen/processor.rs [hunk 17]
```diff
@@ -27,15 +27,15 @@ pub struct OuterQuery<'a, 'b, T: FieldElement> {
...
+    pub connecting_identity: &'a Identity<SelectedExpressions<Expression<T>>>,
     /// The left side of the connecting identity, evaluated.
     pub left: Left<'a, T>,
 }
 
 impl<'a, 'b, T: FieldElement> OuterQuery<'a, 'b, T> {
     pub fn new(
         caller_rows: &'b RowPair<'b, 'a, T>,
-        connecting_identity: &'a Identity<Expression<T>>,
+        connecting_identity: &'a Identity<SelectedExpressions<Expression<T>>>,
```
Comment (chriseth) [inline executor/src/witgen/processor.rs:38]: Same here, all the files in this crate should only need to use `Identity<T>`.

File: parser/tests/parentheses_test.rs [hunk 18]
```diff
@@ -0,0 +1,136 @@
+#[cfg(test)]
```
Comment (chriseth) [inline parser/tests/parentheses_test.rs:1]: I think this file was moved.

File: pil-analyzer/src/evaluator.rs [hunk 19]
```diff
@@ -749,7 +749,12 @@ impl<'a, 'b, T: FieldElement, S: SymbolLookup<'a, T>> Evaluator<'a, 'b, T, S> {
             Expression::BlockExpression(_, BlockExpression { statements, expr }) => {
                 self.op_stack
                     .push(Operation::TruncateLocals(self.local_vars.len()));
-                self.op_stack.push(Operation::Expand(expr));
+                match expr {
+                    Some(expr) => {
+                        self.op_stack.push(Operation::Expand(expr));
+                    }
+                    None => {}
```
Comment (chriseth) [inline pil-analyzer/src/evaluator.rs:756]: Do you have tests covering this line? Shouldn't it at least push an empty tuple or something?

File: pil-analyzer/src/expression_processor.rs [hunk 20]
```diff
@@ -362,3 +381,17 @@ struct LocalVariableState {
     pub local_variables: HashMap<String, u64>,
     pub local_variable_counter: u64,
 }
+
+fn combine_source_refs(refs: Vec<&SourceRef>) -> SourceRef {
```
Comment (chriseth) [inline pil-analyzer/src/expression_processor.rs:385]: Maybe this was not such a good suggestion of mine. What happens if the source refs are from different files? Maybe it's best to keep it at Unknown. Sorry!

File: pil-analyzer/src/pil_analyzer.rs [hunk 21]
```diff
@@ -287,8 +288,18 @@ impl PILAnalyzer {
...
+                            part.expressions.as_mut(),
+                            Type::Array(ArrayType {
+                                base: Box::new(Type::Expr),
+                                length: None,
+                            })
+                            .into(),
+                        ))
+                    } else {
+                        panic!("Selected expressions need to be provided as arrays.\n  Expected [{}].\n  Found: {}",
+                        part.expressions, part.expressions)
```
Comment (chriseth) [inline pil-analyzer/src/pil_analyzer.rs:302]: Formatting looks off.

File: pil-analyzer/src/pil_analyzer.rs [hunk 22]
```diff
@@ -287,8 +288,18 @@ impl PILAnalyzer {
                     if let Some(selector) = &mut part.selector {
                         expressions.push((selector, Type::Expr.into()))
                     }
-                    for e in &mut part.expressions {
-                        expressions.push((e, Type::Expr.into()))
+                    if let Expression::ArrayLiteral(_, _) = part.expressions.as_ref() {
```
Comment (chriseth) [inline pil-analyzer/src/pil_analyzer.rs:291]: Why only for array literals? This should work for any expression. We can now also test things like `x $ f() in g()`.

**Comment by chriseth** [comment]
I think this is almost ready. Sorry for the delay in reviewing. I think this can still be extended, but we can do it later on.

File: pil-analyzer/src/evaluator.rs (hunk 19 above)
Comment (gzanitti) [reply pil-analyzer/src/evaluator.rs:756]: My guess here was that it made no difference to expand an empty expression, hence the empty None case. Also, there is the problem that creating an expression at this point is complex because of the lifetimes.


I'll see what I can do to solve it, thanks for your comment.

File: pil-analyzer/src/pil_analyzer.rs (hunk 22 above)
Comment (gzanitti) [reply pil-analyzer/src/pil_analyzer.rs:291]: I seem to remember that we had talked about this kind of expressions now always being of the form x $ [f()] in [g()], but no problem, I can fix that. The only thing I need to know (if I understand correctly), is how to treat those expressions that are not arrays here:

https://github.com/powdr-labs/powdr/blob/7e3a5c70913537f6351064762eb82ff24abd0199/pil-analyzer/src/condenser.rs#L293

because this function expects &sel_expr.expressions to evaluate to an array (which doesn't happen anymore in the example you gave me).

File: pil-analyzer/src/evaluator.rs (hunk 19 above)
Comment (gzanitti) [reply pil-analyzer/src/evaluator.rs:756]: This test should cover the case you mention, or am I wrong?

https://github.com/powdr-labs/powdr/blob/524d579544eb4d3f2370a5d8af0cb1011ded5546/pil-analyzer/tests/condenser.rs#L208

**Comment by gzanitti** [comment]
> I think this is almost ready. Sorry for the delay in reviewing. I think this can still be extended, but we can do it later on.

You don't have to apologise, it's a bit my fault because it took me a while to get into rhythm and understand a couple of things. It shouldn't happen anymore :)

File: pil-analyzer/src/evaluator.rs (hunk 19 above)
Comment (chriseth) [reply pil-analyzer/src/evaluator.rs:756]: I don't know. Can you put a panic inside there to see if it is hit?

Also can you add a test for the following (also just curious what is happening and if we should disallow it):
```
let x = {};
```

File: pil-analyzer/src/pil_analyzer.rs (hunk 22 above)
Comment (chriseth) [reply pil-analyzer/src/pil_analyzer.rs:291]: Sorry, I should have expanded the example a little: My idea was that `f()` and `g()` evaluate to arrays. So you don't have to support tuples or anything else in `condense_to_array_of_algebraic_expressions`.

What I meant was that you should not expect here in line 291 that part.expressions is syntactically an array literal - only that its type is an array of expr.

File: pil-analyzer/src/evaluator.rs (hunk 19 above)
Comment (chriseth) [reply pil-analyzer/src/evaluator.rs:756]: Oh or maybe also something a bit more complicated:
```
let f = |_| 7;
let g = f({});
```

Comment (gzanitti) [reply pil-analyzer/src/evaluator.rs]: Fixed. To avoid the lifetime issue, what I did was to change the None cases to the empty tuple directly in the statement processing, so at this point, the None case should not be reachable (I added some tests, including this one you mention above).

Comment (chriseth) [reply pil-analyzer/src/evaluator.rs]: Are you sure this is a good solution? I don't really like modifying the AST in such a way. You don't need to deal with the lifetime of expressions here. For an example, see the ::Number case. I think the following should do it:

```
 self.value_stack.push(Value::Tuple(vec![]))
```

File: pil-analyzer/src/condenser.rs [hunk 23]
```diff
@@ -322,7 +321,7 @@ impl<'a, T: FieldElement> Condenser<'a, T> {
                     _ => panic!("Expected expression but got {item}"),
                 })
                 .collect(),
-            _ => panic!("Expected array of algebraic expressions, but got {result}"),
+            _ => panic!("Expected array/tuple of algebraic expressions or an algebraic expression, but got {result}"),
```
Comment (chriseth) [inline pil-analyzer/src/condenser.rs:324]:
```suggestion
            _ => panic!("Expected array of algebraic expressions or an algebraic expression, but got {result}"),
```

File: pil-analyzer/tests/types.rs [hunk 24]
```diff
@@ -516,3 +516,13 @@ fn enum_too_many_fields() {
     ";
     type_check(input, &[]);
 }
+
```
Comment (chriseth) [inline pil-analyzer/tests/types.rs:519]: Can you add another test that checks types for empty match arms, empty if/else blocks and just empty blocks inside fuctions? - or at least blocks with empty expression part.

**Review by chriseth — APPROVED** [review]
